ibis/common/patterns.py

```diff
@@ -24,17 +25,30 @@
 from typing_extensions import GenericMeta, Self, get_args, get_origin
 
 from ibis.common.bases import Singleton, Slotted
-from ibis.common.collections import RewindableIterator, frozendict
+from ibis.common.collections import FrozenDict, RewindableIterator, frozendict
 from ibis.common.dispatch import lazy_singledispatch
 from ibis.common.typing import Sentinel, get_bound_typevars, get_type_params
 from ibis.util import is_iterable, promote_tuple
 
-try:
+if TYPE_CHECKING:
+    from types import ModuleType
+
+if sys.version_info >= (3, 10):
```

Comment:
Can you move these conditional imports (including _ClassInfo below) to `ibis.common.typing` so we place the typing-compatibility related code to a centralized place?